pipeline.py

```diff
@@ -49,6 +49,14 @@ def _execute_one_cycle(self, result_history: list[Result],
                      result_history[-1])
     result_history.append(
         self.writing_stage.execute(result_history=result_history))
+    if (not isinstance(result_history[-1], BuildResult) or
+        not result_history[-1].compiles):
```

Comment:
Why single out "compiles" ? That seems to be tying specific stage logic too much into this Pipeline code. 

Can we instead generalize this some other way to avoid encoding low level logic here? i.e. `result_history[-1].success`. ?